### DISCUSSION TITLE
Read AES128 encryted mp3

### DISCUSSION THREAD

**Discussion by sulsuldev** [schreibfaul1/ESP32-audioI2S#660] [Q&A]
I use the ESP32-audioI2S library on an esp32 under arduino and the output sound quality is really good.
I want to use this library to play AES128 encrypted mp3 file.
But I don't know how to use the library to read files decrypted byte by byte over time.
I don't want to extract the decoded file on the SD card, because it is accessible and the decryption time is too long (1MB->10seconds for extraction). Also, in the long term there are the risk to altered the SD card.
I would like to read the encrypted file on the SD card. In the decrypting loop, I would like to inject the decoded bytes in RAM into the read audio buffer in order to have the sound output. But I don't know how to proceed.
Is it possible to have sample code or ideas to implement it?
Thank you in advance for your feedback.
Regards
Sully

```
#define KEY_AES128 "0123456789abcdef"
#define AES_BLOCK_SIZE 16
mbedtls_aes_context aes;
char *keycrypt = KEY_AES128;

void decryptFileToAudio(const char *encryptedFileName) {
  File encryptedFile = SD.open(encryptedFileName, FILE_READ);

  if (!encryptedFile) {
    Serial.println("Error opening file");
    return;
  }

  mbedtls_aes_init(&aes);
  mbedtls_aes_setkey_dec(&aes, (const unsigned char *)keycrypt, strlen(keycrypt) * 8);  // Taille de clé en bits

  byte bufferIn[AES_BLOCK_SIZE];
  byte bufferOut[AES_BLOCK_SIZE];

  timeDecodeCrypt = millis();
  while (encryptedFile.available()) {

    // Read a block of the encrypted file
    encryptedFile.read(bufferIn, AES_BLOCK_SIZE);

    // Decrypt the block
    mbedtls_aes_crypt_ecb(&aes, MBEDTLS_AES_DECRYPT, bufferIn, bufferOut);

    //TO DO
    //HOW TO LOAD the decrypted block "bufferOut" on the audio buffer chunk by chunk
    //I DON'T KNOW HOW TO LOAD DIRECTLY THE BUFFER OF THE AUDIO LIBRARIE
    //?? audio.connecttohost(bufferOut);      
  }
  mbedtls_aes_free(&aes);
  encryptedFile.close();
}
```

**Comment by schreibfaul1** [comment]
One possibility would be to decrypt the data read from the SD card in `processLocalFile()` before it is written to the audio buffer.
````c++
 int32_t bytesAddedToBuffer = audiofile.read(InBuff.getWritePtr(), availableBytes);
````

**Reply by sulsuldev** [reply]
Thank you for your answer, but I don't see how to write into the audio buffer directly.

In my decryption loop of the decryptFileToAudio("file_encrypt.enc") function which is outside the I2Saudio function, the extraction of encrypted SD files is done in packets of 128 bits (16 bytes). Then, I don't see how to transfer the 128-bit packets to the audio buffer in the processLocalFile() function.

Do I need to modify the processLocalFile() function? If yes, what modification should I make to the line "int32_t bytesAddedToBuffer = audiofile.read(InBuff.getWritePtr(), availableBytes);" to load the different 16-byte packets that are decrypted.

Additionally, I don't see how to initiate the loading of the processLocalFile() function, informing the decoding type as mp3. In the generic call, we launch the function "audio.connecttoFS(SD, "test.mp3");", the *.mp3 indicates to the function the audio decoding type. If I make a call with the encrypted file with the command "audio.connecttoFS(SD, "test_enc.mp3");" the mp3 header will not be visible directly because the file is encrypted.

Thanks in advance for your feedback.
Regards
Sully

**Reply by schreibfaul1** [reply]
If you create a buffer for the decrypted data (e.g. ~4KB in size) then the line could look like this:
`
memcpy(InBuff.getWritePtr(), decrypredData, nrOfDecryptedBytes);
int32_t bytesAddedToBuffer = nrOfDecryptedBytes;
`

**Reply by sulsuldev** [reply]
Thank you for your answer.  

### First modification:
I tried to decode directly in the processLocalFile function when using a single 128 bit packet.  I don't have any log info appearing that shows there are problems.  However, the sound is choppy.  it's as if the buffer was not filled.  how much data is needed in the audio buffer?

In my main program, i call the function like this : 
```
audio.connecttoFS(SD, "/medium_file_encrypt.mp3");
```

In the function processLocalFile
I replace :
```
int32_t bytesAddedToBuffer = audiofile.read(InBuff.getWritePtr(), availableBytes);
```
by theses lines :
```
#ifdef _MODIF_SULSULDEV

	// Decrypt on packet of 128bits and injection on the AudioBuffer

	uint32_t nrOfDecryptedBytes=16;
	unsigned char encrypredData[16]={0};
	
	//read audio file
	audiofile.read(encrypredData, nrOfDecryptedBytes);

	// decrypt by packet of 128bits
	mbedtls_aes_crypt_ecb(&aes, MBEDTLS_AES_DECRYPT, encrypredData, decrypredData);
	
	//buffer transfert to audio buffer
	memcpy(InBuff.getWritePtr(), decrypredData, nrOfDecryptedBytes); 
	
	int32_t bytesAddedToBuffer = nrOfDecryptedBytes;
	
#else

	int32_t bytesAddedToBuffer = audiofile.read(InBuff.getWritePtr(), availableBytes);
	
#endif
```

### Second modification
In the processLocalFile function, I tried to add a loop to have a larger buffer of several 128-bit packets.  I have no sound output and I have infos log errors "mp3 decode error" and "sync word found info xxx"

Modification of the processLocalFile function for testing more datas in the audio buffer
```
#ifdef _MODIF_SULSULDEV

	// Decrypt 128bits and injection on the AudioBuffer
	int intpacket128bits=10;
	int32_t bytesAddedToBuffer =0;
	uint32_t nrOfDecryptedBytes=16;
	unsigned char encrypredData[16]={0};
	
	for (int cc=0;cc<intpacket128bits;cc++)
	{
		//read audio file
		audiofile.read(encrypredData, nrOfDecryptedBytes);

		// decrypt by packet of 128bits
		mbedtls_aes_crypt_ecb(&aes, MBEDTLS_AES_DECRYPT, encrypredData, decrypredData);
		
		//buffer transfert to audio buffer
		memcpy(InBuff.getWritePtr(), decrypredData, nrOfDecryptedBytes); 
	}
	
	int32_t bytesAddedToBuffer = nrOfDecryptedBytes*intpacket128bits;
	
#else

	int32_t bytesAddedToBuffer = audiofile.read(InBuff.getWritePtr(), availableBytes);
	
#endif
```

I think I'm using the function wrong and I don't know how to call an external function. But I don't see how to do it.

### Third modification:
I tried to do the decoding outside the function in the main program but I don't know how to start or call the reading audio function and how to inject a decryted buffer into the into the audio bufer InBuff in the processLocalFile function.
In the main function, I read the encrypted file and I decrypt it to load the decrypted buffer but I do not see how to launch the audio in parallel, because I cannot call the same file with the function audio.connecttoFS("file").

Thanks in advance for your feedback.
Regards
Sully
